test/integration/banListTest.ts

```diff
@@ -225,3 +227,63 @@ describe('Test: We will not be able to ban ourselves via ACL.', function () {
+        const serverName: string = new UserID(await mjolnir.getUserId()).domain
+        const moderator = await newTestUser({ name: { contains: "moderator" }});
+        moderator.joinRoom(this.mjolnir.managementRoomId);
+        const mjolnirId = await mjolnir.getUserId();
+        // How many rooms are we going to protect? Do we need to protect a few and verify all the ACL updates were batched in each?
+        const protectedRooms: string[] = [];
+        for (let i = 0; i < 10; i++) {
+            const room = await moderator.createRoom({ invite: [mjolnirId]});
+            await mjolnir.joinRoom(room);
+            await moderator.setUserPowerLevel(mjolnirId, room, 100);
+            await this.mjolnir!.addProtectedRoom(room);
+            protectedRooms.push(room);
+        }
+
+        // If a previous test hasn't cleaned up properly, these rooms will be populated by bogus ACLs at this point.
+        await this.mjolnir!.syncLists();
+        await Promise.all(protectedRooms.map(async room => {
+            // We're going to need timeline pagination I'm afraid.
+            const roomAcl = await mjolnir.getRoomStateEvent(room, "m.room.server_acl", "");
+            assert.equal(roomAcl?.deny?.length ?? 0, 0, 'There should be no entries in the deny ACL.');
+        }));
+
+        // Flood the subsribed list with banned servers, which should prompt Mjolnir to update server ACL in protected rooms.
+        const banListId = await moderator.createRoom({ invite: [mjolnirId] });
+        mjolnir.joinRoom(banListId);
+        this.mjolnir!.watchList(Permalinks.forRoom(banListId));
+        const acl = new ServerAcl(serverName).denyIpAddresses().allowServer("*");
+        for (let i = 0; i < 200; i++) {
+            acl.denyServer(i.toString());
+            await createPolicyRule(moderator, banListId, RULE_SERVER, `${i}`, `Rule #${i}`);
+            // Give them a bit of a spread over time.
+            await new Promise(resolve => setTimeout(resolve, 5));
+        }
+
+        // We do this because it should force us to wait until all the ACL events have been applied.
+        await this.mjolnir!.syncLists();
+
+        // Check each of the protected rooms.
+        await Promise.all(protectedRooms.map(async room => {
+            const roomAcl = await mjolnir.getRoomStateEvent(room, "m.room.server_acl", "");
+            if (!acl.matches(roomAcl)) {
+                assert.fail(`Room ${room} doesn't have the correct ACL: ${JSON.stringify(roomAcl, null, 2)}`)
+            }
+            let aclEventCount = 0;
+            await getMessagesByUserIn(mjolnir, mjolnirId, room, 100, events => {
+                events.forEach(event => event.type === 'm.room.server_acl' ? aclEventCount += 1 : null);
+            });
+            LogService.debug('BanListTest', `aclEventCount: ${aclEventCount}`);
+            // If there's less than two then it probably means the ACL was put in by this test calling syncLists
```

Comment:
Instead of testing that it's in [3, 9], could you check that there is only one event of type `m.room.server_acl` that was sent since the start of the loop that calls `createPolicyRule`? Would that work? If so, it feels like it would be more robust.

Reply:
I don't know how well this test is going to run on CI or other people's environments, which yes is why you're asking this too. But the issue is the margin between triggering the 3s batch and not triggering it is difficult to pin down in a way that works everywhere and it does need to be tested too (so i think it would be hard to have a test where we trigger it it just once). So having a wider range [3, 9] seemed to be better in this regard even if it is more arbitrary.